File: android/src/main/java/com/example/voicevideocall/VoiceVideoCallPlugin.java
package com.example.voicevideocall;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.webrtc.*;
import com.onesignal.OneSignal;

@CapacitorPlugin(name = "VoiceVideoCall")
public class VoiceVideoCallPlugin extends Plugin {
    private static final String TAG = "VoiceVideoCallPlugin";
    private PeerConnectionFactory peerConnectionFactory;
    private PeerConnection peerConnection;
    private MediaStream mediaStream;
    private VideoTrack localVideoTrack;
    private AudioTrack localAudioTrack;

    @Override
    public void load() {
        // Initialize WebRTC
        PeerConnectionFactory.InitializationOptions initializationOptions =
                PeerConnectionFactory.InitializationOptions.builder(getContext())
                        .setEnableInternalTracer(true)
                        .createInitializationOptions();
        PeerConnectionFactory.initialize(initializationOptions);

        // Create PeerConnectionFactory
        PeerConnectionFactory.Options options = new PeerConnectionFactory.Options();
        peerConnectionFactory = PeerConnectionFactory.builder()
                .setOptions(options)
                .createPeerConnectionFactory();

        // Initialize OneSignal
        OneSignal.setLogLevel(OneSignal.LOG_LEVEL.VERBOSE, OneSignal.LOG_LEVEL.NONE);
        OneSignal.initWithContext(getContext());
    }

    @PluginMethod
    public void startCall(PluginCall call) {
        String roomId = call.getString("roomId");
        String callType = call.getString("callType");

        if (roomId == null || callType == null) {
            call.reject("Room ID and call type are required");
            return;
        }

        try {
            // Create media constraints
            MediaConstraints mediaConstraints = new MediaConstraints();
            mediaConstraints.mandatory.add(
                    new MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"));
            mediaConstraints.mandatory.add(
                    new MediaConstraints.KeyValuePair("OfferToReceiveVideo", 
                            callType.equals("video") ? "true" : "false"));

            // Create peer connection
            PeerConnection.RTCConfiguration rtcConfig = 
                    new PeerConnection.RTCConfiguration(new ArrayList<>());
            peerConnection = peerConnectionFactory.createPeerConnection(
                    rtcConfig, new PeerConnection.Observer() {
                @Override
                public void onSignalingChange(PeerConnection.SignalingState signalingState) {
                    Log.d(TAG, "onSignalingChange: " + signalingState);
                }

                @Override
                public void onIceConnectionChange(PeerConnection.IceConnectionState iceConnectionState) {
                    Log.d(TAG, "onIceConnectionChange: " + iceConnectionState);
                }

                @Override
                public void onIceConnectionReceivingChange(boolean b) {
                    Log.d(TAG, "onIceConnectionReceivingChange: " + b);
                }

                @Override
                public void onIceGatheringChange(PeerConnection.IceGatheringState iceGatheringState) {
                    Log.d(TAG, "onIceGatheringChange: " + iceGatheringState);
                }

                @Override
                public void onIceCandidate(IceCandidate iceCandidate) {
                    // Send ice candidate through OneSignal
                    JSObject candidateData = new JSObject();
                    candidateData.put("type", "candidate");
                    candidateData.put("candidate", iceCandidate.sdp);
                    candidateData.put("sdpMLineIndex", iceCandidate.sdpMLineIndex);
                    candidateData.put("sdpMid", iceCandidate.sdpMid);
                    
                    OneSignal.postNotification(candidateData.toString(),
                            null,
                            roomId);
                }

                @Override
                public void onIceCandidatesRemoved(IceCandidate[] iceCandidates) {
                    Log.d(TAG, "onIceCandidatesRemoved");
                }

                @Override
                public void onAddStream(MediaStream mediaStream) {
                    Log.d(TAG, "onAddStream");
                }

                @Override
                public void onRemoveStream(MediaStream mediaStream) {
                    Log.d(TAG, "onRemoveStream");
                }

                @Override
                public void onDataChannel(DataChannel dataChannel) {
                    Log.d(TAG, "onDataChannel");
                }

                @Override
                public void onRenegotiationNeeded() {
                    Log.d(TAG, "onRenegotiationNeeded");
                }
            });

            // Create local media stream
            mediaStream = peerConnectionFactory.createLocalMediaStream("ARDAMS");
            
            // Add audio track
            AudioSource audioSource = peerConnectionFactory.createAudioSource(new MediaConstraints());
            localAudioTrack = peerConnectionFactory.createAudioTrack("ARDAMSa0", audioSource);
            mediaStream.addTrack(localAudioTrack);

            // Add video track if video call
            if (callType.equals("video")) {
                VideoSource videoSource = peerConnectionFactory.createVideoSource(false);
                localVideoTrack = peerConnectionFactory.createVideoTrack("ARDAMSv0", videoSource);
                mediaStream.addTrack(localVideoTrack);
            }

            // Add stream to peer connection
            peerConnection.addStream(mediaStream);

            // Create offer
            peerConnection.createOffer(new SdpObserver() {
                @Override
                public void onCreateSuccess(SessionDescription sessionDescription) {
                    peerConnection.setLocalDescription(new SdpObserver() {
                        @Override
                        public void onCreateSuccess(SessionDescription sessionDescription) {}

                        @Override
                        public void onSetSuccess() {
                            // Send offer through OneSignal
                            JSObject offerData = new JSObject();
                            offerData.put("type", "offer");
                            offerData.put("sdp", sessionDescription.description);
                            
                            OneSignal.postNotification(offerData.toString(),
                                    null,
                                    roomId);
                        }

                        @Override
                        public void onCreateFailure(String s) {
                            call.reject("Failed to create local description: " + s);
                        }

                        @Override
                        public void onSetFailure(String s) {
                            call.reject("Failed to set local description: " + s);
                        }
                    }, sessionDescription);
                }

                @Override
                public void onSetSuccess() {}

                @Override
                public void onCreateFailure(String s) {
                    call.reject("Failed to create offer: " + s);
                }

                @Override
                public void onSetFailure(String s) {}
            }, mediaConstraints);

            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to start call: " + e.getMessage());
        }
    }

    @PluginMethod
    public void acceptCall(PluginCall call) {
        String roomId = call.getString("roomId");
        if (roomId == null) {
            call.reject("Room ID is required");
            return;
        }

        try {
            // Implementation similar to startCall but handles incoming call
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to accept call: " + e.getMessage());
        }
    }

    @PluginMethod
    public void rejectCall(PluginCall call) {
        String roomId = call.getString("roomId");
        if (roomId == null) {
            call.reject("Room ID is required");
            return;
        }

        try {
            // Send rejection notification through OneSignal
            JSObject rejectData = new JSObject();
            rejectData.put("type", "reject");
            
            OneSignal.postNotification(rejectData.toString(),
                    null,
                    roomId);
            
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to reject call: " + e.getMessage());
        }
    }

    @PluginMethod
    public void endCall(PluginCall call) {
        try {
            if (peerConnection != null) {
                peerConnection.close();
                peerConnection = null;
            }
            
            if (mediaStream != null) {
                mediaStream.dispose();
                mediaStream = null;
            }
            
            if (localVideoTrack != null) {
                localVideoTrack.dispose();
                localVideoTrack = null;
            }
            
            if (localAudioTrack != null) {
                localAudioTrack.dispose();
                localAudioTrack = null;
            }
            
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to end call: " + e.getMessage());
        }
    }
}

File: ios/Plugin/VoiceVideoCallPlugin.swift
import Foundation
import Capacitor
import WebRTC
import OneSignal

@objc(VoiceVideoCallPlugin)
public class VoiceVideoCallPlugin: CAPPlugin {
    private var peerConnectionFactory: RTCPeerConnectionFactory?
    private var peerConnection: RTCPeerConnection?
    private var mediaStream: RTCMediaStream?
    private var localVideoTrack: RTCVideoTrack?
    private var localAudioTrack: RTCAudioTrack?
    
    override public func load() {
        // Initialize WebRTC
        RTCInitializeSSL()
        let videoEncoderFactory = RTCDefaultVideoEncoderFactory()
        let videoDecoderFactory = RTCDefaultVideoDecoderFactory()
        peerConnectionFactory = RTCPeerConnectionFactory(
            encoderFactory: videoEncoderFactory,
            decoderFactory: videoDecoderFactory
        )
        
        // Initialize OneSignal
        OneSignal.setLogLevel(.LL_VERBOSE, visualLevel: .LL_NONE)
    }
    
    @objc func startCall(_ call: CAPPluginCall) {
        guard let roomId = call.getString("roomId"),
              let callType = call.getString("callType") else {
            call.reject("Room ID and call type are required")
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                call.reject("Plugin instance is not available")
                return
            }
            
            do {
                // Configure peer connection
                let config = RTCConfiguration()
                config.iceServers = []
                
                // Media constraints
                let constraints = RTCMediaConstraints(
                    mandatoryConstraints: [
                        "OfferToReceiveAudio": "true",
                        "OfferToReceiveVideo": callType == "video" ? "true" : "false"
                    ],
                    optionalConstraints: nil
                )
                
                // Create peer connection
                self.peerConnection = self.peerConnectionFactory?.peerConnection(
                    with: config,
                    constraints: constraints,
                    delegate: self
                )
                
                // Create media stream
                guard let mediaStream = self.peerConnectionFactory?.mediaStream(withStreamId: "ARDAMS") else {
                    call.reject("Failed to create media stream")
                    return
                }
                self.mediaStream = mediaStream
                
                // Add audio track
                let audioConstraints = RTCMediaConstraints(
                    mandatoryConstraints: nil,
                    optionalConstraints: nil
                )
                guard let audioSource = self.peerConnectionFactory?.audioSource(with: audioConstraints),
                      let audioTrack = self.peerConnectionFactory?.audioTrack(
                        with: audioSource,
                        trackId: "ARDAMSa0"
                      ) else {
                    call.reject("Failed to create audio track")
                    return
                }
                self.localAudioTrack = audioTrack
                mediaStream.addAudioTrack(audioTrack)
                
                // Add video track if video call
                if callType == "video" {
                    guard let videoSource = self.peerConnectionFactory?.videoSource(),
                          let videoTrack = self.peerConnectionFactory?.videoTrack(
                            with: videoSource,
                            trackId: "ARDAMSv0"
                          ) else {
                        call.reject("Failed to create video track")
                        return
                    }
                    self.localVideoTrack = videoTrack
                    mediaStream.addVideoTrack(videoTrack)
                }
                
                // Add stream to peer connection
                self.peerConnection?.add(mediaStream)
                
                // Create offer
                self.peerConnection?.offer(for: constraints) { [weak self] (sessionDescription, error) in
                    guard let self = self else { return }
                    
                    if let error = error {
                        call.reject("Failed to create offer: \(error.localizedDescription)")
                        return
                    }
                    
                    guard let sessionDescription = sessionDescription else {
                        call.reject("Failed to create session description")
                        return
                    }
                    
                    self.peerConnection?.setLocalDescription(sessionDescription) { error in
                        if let error = error {
                            call.reject("Failed to set local description: \(error.localizedDescription)")
                            return
                        }
                        
                        // Send offer through OneSignal
                        let offerData: [String: Any] = [
                            "type": "offer",
                            "sdp": sessionDescription.sdp
                        ]
                        
                        OneSignal.postNotification(
                            ["contents": offerData],
                            toUserId: roomId,
                            onSuccess: { _ in
                                call.resolve()
                            },
                            onFailure: { error in
                                call.reject("Failed to send offer: \(error)")
                            }
                        )
                    }
                }
            } catch {
                call.reject("Failed to start call: \(error.localizedDescription)")
            }
        }
    }
    
    @objc func acceptCall(_ call: CAPPluginCall) {
        guard let roomId = call.getString("roomId") else {
            call.reject("Room ID is required")
            return
        }
        
        // Implementation similar to startCall but handles incoming call
        call.resolve()
    }
    
    @objc func rejectCall(_ call: CAPPluginCall) {
        guard let roomId = call.getString("roomId") else {
            call.reject("Room ID is required")
            return
        }
        
        // Send rejection notification through OneSignal
        let rejectData: [String: Any] = [
            "type": "reject"
        ]
        
        OneSignal.postNotification(
            ["contents": rejectData],
            toUserId: roomId,
            onSuccess: { _ in
                call.resolve()
            },
            onFailure: { error in
                call.reject("Failed to send rejection: \(error)")
            }
        )
    }
    
    @objc func endCall(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                call.reject("Plugin instance is not available")
                return
            }
            
            self.peerConnection?.close()
            self.peerConnection = nil
            
            self.mediaStream = nil
            self.localVideoTrack = nil
            self.localAudioTrack = nil
            
            call.resolve()
        }
    }
}

// MARK: - RTCPeerConnectionDelegate
extension VoiceVideoCallPlugin: RTCPeerConnectionDelegate {
    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
        print("Signaling state changed: \(stateChanged)")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        print("Stream added")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
        print("Stream removed")
    }
    
    public func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
        print("Should negotiate")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        print("ICE connection state changed: \(newState)")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {
        print("ICE gathering state changed: \(newState)")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        // Send ice candidate through OneSignal
        let candidateData: [String: Any] = [
            "type": "candidate",
            "candidate": candidate.sdp,
            "sdpMLineIndex": candidate.sdpMLineIndex,
            "sdpMid": candidate.sdpMid ?? ""
        ]
        
        // Note: You'll need to implement the actual sending logic here
        print("Ice candidate generated: \(candidateData)")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {
        print("Ice candidates removed")
    }
    
    public func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        print("Data channel opened")
    }
}